evm/src/cpu/kernel/asm/sha2/ops.asm

```diff
@@ -0,0 +1,189 @@
+// u32 addition (discarding 2^32 bit)
+%macro add_u32
+    // stack: x, y
+    ADD
+    // stack: x + y
+    DUP1
+    // stack: x + y, x + y
+    %shr_const(32)
+    // stack: (x + y) >> 32, x + y
+    %shl_const(32)
+    // stack: ((x + y) >> 32) << 32, x + y
+    SWAP1
+    // stack: x + y, ((x + y) >> 32) << 32
+    SUB
+    // stack: x + y - ((x + y) >> 32) << 32
+%endmacro
+
+
+// 32-bit right rotation
+%macro rotr
```

Comment:
I think all the callers pass a constant `rot`, so I'd suggest making `rot` a macro var which should simplify things a bit.